### ISSUE TITLE
Flysystem "delete" AWS-S3 exception

### ISSUE THREAD

**Issue by niden** [thephpleague/flysystem#1030]
### Question
Hello everyone

We have been using Flysystem for local files for some months now and it works perfectly. We recently embarked on changing our storage to S3.

We updated composer with the relevant pacakges required

```
    "aws/aws-sdk-php": "^3.54",
    "league/flysystem": "^1.0",
    "league/flysystem-aws-s3-v3": "^1.0",
```
Making additional changes to define the bucket, secret and key we are in business. Whenever we upload a file, it gets stored in the bucket perfectly. The path we use is something like this

```
data/m228/2019/05/31/79428b23ae7d7eb779d607d428e9aca1b140506d
```
We can read the file (using the library and our application) and it is also visible in the S3 AWS console.

The issue we have is when we try to delete the file. Issuing:

```php
$this->adapter->delete($this->getFileName());
```
results in a 404 exception like this:
```txt
https://test-test.s3.amazonaws.com/data/m228/2019/05/31/79428b23ae7d7eb779d607d428e9aca1b140506d` resulted in a `404 Not Found` response in /app/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php on line 113
```
I noticed in this issue https://github.com/thephpleague/flysystem/issues/477 that someone had a similar experience but their issue was that they were passing the whole URL and not the absolute path of the file. I have confirmed that the `getFileName()` above is indeed:
```
data/m228/2019/05/31/79428b23ae7d7eb779d607d428e9aca1b140506d
```

Note the file **is** deleted from the bucket successfully.

Also tried to wrap the code in a `try`/`catch` but that did not work since the exception comes from Guzzle.

Any pointers are more than welcome

**Comment by frankdejonge** [comment]
Yassas Nikolaos,

I just created this sample script to try and reproduce your situation but for me it's working fine:

```php
<?php

use Aws\S3\S3Client;
use League\Flysystem\AwsS3v3\AwsS3Adapter;
use League\Flysystem\Filesystem;

include __DIR__.'/vendor/autoload.php';

$client = new S3Client([
    'version' => 'latest',
    'credentials' => [
        'key' => 'redacted' 
        'secret' => 'redacted',
    ],
    'region' => 'eu-central-1',
]);
$adapter = new AwsS3Adapter($client, 'some-bucket');
$filesystem = new Filesystem($adapter);
$filesystem->write('data/m228/2019/05/31/79428b23ae7d7eb779d607d428e9aca1b140506d', 'data');
var_dump($filesystem->delete('data/m228/2019/05/31/79428b23ae7d7eb779d607d428e9aca1b140506d'));
```

In case of an invalid path you should get a different type of exception, a flysystem specific one. Perhaps you can paste a stack trace of the issue here and I can more clearly determine what's going on.

For now my advice would be to check the stack trace. Also, double check the configuration of the adapter(s?), check for any inconsistencies. And try updating to the latest package of all the related dependencies. For now I don't have more advice due to the amount of information provided.

Tomorrow I'll be going on holiday so my response might take a little longer than usual.

**Comment by niden** [comment]
@frankdejonge Thank you for the quick response. The only difference I can see right now is that you use the method `write` to store the file while we use `put`. Other than that the scripts are pretty much identical.

I will check the stack trace and if necessary post it here with additional information over the weekend.

Thank you!

**Comment by frankdejonge** [comment]
@niden that call shouldn't matter, but just to be sure I just verified the same behaviour with the `write` and `put` method.

**Comment by niden** [comment]
So I got the full stack and it appears that this thing happens from a `HEAD` request (sorry for the length of the stack).

I confirmed once again that the sample you posted is identical (apart from the `put` vs `write) to our code. The only difference is that we use a `StorageTrait` where we attach the file related methods to any part of our application but in all the CRUD for files is the same, wrapping the Flysystem.

```
GuzzleHttp\Exception\ClientException: Client error: 
`HEAD https://test-team.s3.amazonaws.com/data/m228/2019/06/01/1708092c8dfb526ea445452eb1be25357e95e2c5` resulted in a 
`404 Not Found` response in /app/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php on line 113

Call Stack:
    0.2006     383568   1. {main}() /app/apps/web/public/index.php:0
    0.2593    2669696   2. TDM\Web\Bootstrap\Web->run() /app/apps/web/public/index.php:10
    0.2593    2669696   3. Phalcon\Mvc\Application->handle() /app/apps/web/library/Bootstrap/Web.php:41
    0.2602    2697624   4. Phalcon\Mvc\Dispatcher->dispatch() /app/apps/web/library/Bootstrap/Web.php:41
    0.4398    4729800   5. Phalcon\Mvc\Dispatcher->callActionMethod() /app/apps/web/library/Bootstrap/Web.php:41
    0.4398    4730184   6. TDM\Web\Training\Controllers\DataController->deleteAction() /app/apps/web/library/Bootstrap/Web.php:41
    0.4940    4916080   7. TDM\Web\DataStorageManager->deleteFile() /app/apps/web/app/modules/training/controllers/DataController.php:366
    0.4940    4916080   8. TDM\Web\DataStorageManager->fileDelete() /app/apps/web/library/DataStorageManager.php:45
    0.5047    5023552   9. TDM\Storage->delete() /app/library/Traits/StorageTrait.php:281
    0.6903    9339168  10. League\Flysystem\Filesystem->delete() /app/library/Storage.php:42
    0.7268    9348480  11. League\Flysystem\AwsS3v3\AwsS3Adapter->delete() /app/vendor/league/flysystem/src/Filesystem.php:237
    0.9617    9363208  12. League\Flysystem\AwsS3v3\AwsS3Adapter->has() /app/vendor/league/flysystem-aws-s3-v3/src/AwsS3Adapter.php:188
    0.9618    9363304  13. Aws\S3\S3Client->doesObjectExist() /app/vendor/league/flysystem-aws-s3-v3/src/AwsS3Adapter.php:234
    0.9618    9363904  14. Aws\S3\S3Client->checkExistenceWithCommand() /app/vendor/aws/aws-sdk-php/src/S3/S3ClientTrait.php:272
    0.9619    9363904  15. Aws\S3\S3Client->execute() /app/vendor/aws/aws-sdk-php/src/S3/S3ClientTrait.php:287
    0.9665    9381632  16. GuzzleHttp\Promise\Promise->wait() /app/vendor/aws/aws-sdk-php/src/AwsClientTrait.php:58
    0.9665    9381632  17. GuzzleHttp\Promise\Promise->waitIfPending() /app/vendor/guzzlehttp/promises/src/Promise.php:62
    0.9665    9381632  18. GuzzleHttp\Promise\Promise->invokeWaitList() /app/vendor/guzzlehttp/promises/src/Promise.php:225
    0.9769    9409152  19. GuzzleHttp\Promise\Promise->waitIfPending() /app/vendor/guzzlehttp/promises/src/Promise.php:267
    0.9769    9409152  20. GuzzleHttp\Promise\Promise->invokeWaitList() /app/vendor/guzzlehttp/promises/src/Promise.php:225
    0.9769    9409152  21. GuzzleHttp\Promise\Promise->waitIfPending() /app/vendor/guzzlehttp/promises/src/Promise.php:267
    0.9769    9409152  22. GuzzleHttp\Promise\Promise->invokeWaitFn() /app/vendor/guzzlehttp/promises/src/Promise.php:223
    0.9769    9409152  23. GuzzleHttp\Handler\CurlMultiHandler->execute() /app/vendor/guzzlehttp/promises/src/Promise.php:246
    0.9928    9413000  24. GuzzleHttp\Handler\CurlMultiHandler->tick() /app/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:125
    0.9928    9413000  25. GuzzleHttp\Promise\TaskQueue->run() /app/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:98
    0.9930    9408840  26. GuzzleHttp\Promise\Promise::GuzzleHttp\Promise\{closure}() /app/vendor/guzzlehttp/promises/src/TaskQueue.php:47
    0.9931    9408840  27. GuzzleHttp\Promise\Promise::callHandler() /app/vendor/guzzlehttp/promises/src/Promise.php:156
    0.9931    9408840  28. GuzzleHttp\Middleware::GuzzleHttp\{closure}() /app/vendor/guzzlehttp/promises/src/Promise.php:203


Variables in local scope (#28):
  $code = 404
  $handler = class GuzzleHttp\RedirectMiddleware { private $nextHandler = class Closure { public $static = array (...); public $parameter = array (...) } }
  $request = class GuzzleHttp\Psr7\Request { private $method = 'HEAD'; private $requestTarget = NULL; private $uri = class GuzzleHttp\Psr7\Uri { private $scheme = 'https'; private $userInfo = ''; private $host = 'test-team.s3.amazonaws.com'; private $port = NULL; private $path = '/data/m228/2019/06/01/1708092c8dfb526ea445452eb1be25357e95e2c5'; private $query = ''; private $fragment = '' }; private $headers = array ('Host' => array (...), 'aws-sdk-invocation-id' => array (...), 'aws-sdk-retry' => array (...), 'X-Amz-Content-Sha256' => array (...), 'X-Amz-Date' => array (...), 'Authorization' => array (...), 'User-Agent' => array (...)); private $headerNames = array ('host' => 'Host', 'aws-sdk-invocation-id' => 'aws-sdk-invocation-id', 'aws-sdk-retry' => 'aws-sdk-retry', 'x-amz-content-sha256' => 'X-Amz-Content-Sha256', 'x-amz-date' => 'X-Amz-Date', 'authorization' => 'Authorization', 'user-agent' => 'User-Agent'); private $protocol = '1.1'; private $stream = class GuzzleHttp\Psr7\Stream { private $stream = resource(272) of type (stream); private $size = 0; private $seekable = TRUE; private $readable = TRUE; private $writable = TRUE; private $uri = 'php://temp'; private $customMetadata = array (...) } }
  $response = class GuzzleHttp\Psr7\Response { private $reasonPhrase = 'Not Found'; private $statusCode = 404; private $headers = array ('x-amz-request-id' => array (...), 'x-amz-id-2' => array (...), 'Content-Type' => array (...), 'Transfer-Encoding' => array (...), 'Date' => array (...), 'Server' => array (...)); private $headerNames = array ('x-amz-request-id' => 'x-amz-request-id', 'x-amz-id-2' => 'x-amz-id-2', 'content-type' => 'Content-Type', 'transfer-encoding' => 'Transfer-Encoding', 'date' => 'Date', 'server' => 'Server'); private $protocol = '1.1'; private $stream = class GuzzleHttp\Psr7\Stream { private $stream = resource(275) of type (stream); private $size = 0; private $seekable = TRUE; private $readable = TRUE; private $writable = TRUE; private $uri = 'php://temp'; private $customMetadata = array (...) } }


Aws\S3\Exception\S3Exception: Error executing "HeadObject" on 
"https://test-team.s3.amazonaws.com/data/m228/2019/06/01/1708092c8dfb526ea445452eb1be25357e95e2c5"; 
AWS HTTP error: Client error: `HEAD https://test-team.s3.amazonaws.com/data/m228/2019/06/01/1708092c8dfb526ea445452eb1be25357e95e2c5` resulted in a 
`404 Not Found` response NotFound (client): 404 Not Found (Request-ID: A36CB56A0CD87578) -  in /app/vendor/aws/aws-sdk-php/src/WrappedHttpHandler.php on line 191

Call Stack:
    0.2006     383568   1. {main}() /app/apps/web/public/index.php:0
    0.2593    2669696   2. TDM\Web\Bootstrap\Web->run() /app/apps/web/public/index.php:10
    0.2593    2669696   3. Phalcon\Mvc\Application->handle() /app/apps/web/library/Bootstrap/Web.php:41
    0.2602    2697624   4. Phalcon\Mvc\Dispatcher->dispatch() /app/apps/web/library/Bootstrap/Web.php:41
    0.4398    4729800   5. Phalcon\Mvc\Dispatcher->callActionMethod() /app/apps/web/library/Bootstrap/Web.php:41
    0.4398    4730184   6. TDM\Web\Training\Controllers\DataController->deleteAction() /app/apps/web/library/Bootstrap/Web.php:41
    0.4940    4916080   7. TDM\Web\DataStorageManager->deleteFile() /app/apps/web/app/modules/training/controllers/DataController.php:366
    0.4940    4916080   8. TDM\Web\DataStorageManager->fileDelete() /app/apps/web/library/DataStorageManager.php:45
    0.5047    5023552   9. TDM\Storage->delete() /app/library/Traits/StorageTrait.php:281
    0.6903    9339168  10. League\Flysystem\Filesystem->delete() /app/library/Storage.php:42
    0.7268    9348480  11. League\Flysystem\AwsS3v3\AwsS3Adapter->delete() /app/vendor/league/flysystem/src/Filesystem.php:237
    0.9617    9363208  12. League\Flysystem\AwsS3v3\AwsS3Adapter->has() /app/vendor/league/flysystem-aws-s3-v3/src/AwsS3Adapter.php:188
    0.9618    9363304  13. Aws\S3\S3Client->doesObjectExist() /app/vendor/league/flysystem-aws-s3-v3/src/AwsS3Adapter.php:234
    0.9618    9363904  14. Aws\S3\S3Client->checkExistenceWithCommand() /app/vendor/aws/aws-sdk-php/src/S3/S3ClientTrait.php:272
    0.9619    9363904  15. Aws\S3\S3Client->execute() /app/vendor/aws/aws-sdk-php/src/S3/S3ClientTrait.php:287
    0.9665    9381632  16. GuzzleHttp\Promise\Promise->wait() /app/vendor/aws/aws-sdk-php/src/AwsClientTrait.php:58


Variables in local scope (#16):
  $inner = class Aws\S3\Exception\S3Exception { private ${Aws\Exception\AwsException}response = class GuzzleHttp\Psr7\Response { private $reasonPhrase = 'Not Found'; private $statusCode = 404; private $headers = array (...); private $headerNames = array (...); private $protocol = '1.1'; private $stream = class GuzzleHttp\Psr7\Stream { ... } }; private ${Aws\Exception\AwsException}request = class GuzzleHttp\Psr7\Request { private $method = 'HEAD'; private $requestTarget = NULL; private $uri = class GuzzleHttp\Psr7\Uri { ... }; private $headers = array (...); private $headerNames = array (...); private $protocol = '1.1'; private $stream = class GuzzleHttp\Psr7\Stream { ... } }; private ${Aws\Exception\AwsException}result = NULL; private ${Aws\Exception\AwsException}command = class Aws\Command { private $name = 'HeadObject'; private $handlerList = class Aws\HandlerList { ... }; private $data = array (...) }; private ${Aws\Exception\AwsException}requestId = 'A36CB56A0CD87578'; private ${Aws\Exception\AwsException}errorType = 'client'; private ${Aws\Exception\AwsException}errorCode = 'NotFound'; private ${Aws\Exception\AwsException}connectionError = FALSE; private ${Aws\Exception\AwsException}transferInfo = array ('http' => array (...)); private ${Aws\Exception\AwsException}errorMessage = '404 Not Found (Request-ID: A36CB56A0CD87578)'; private ${Aws\Exception\AwsException}maxRetriesExceeded = FALSE; protected $message = 'Error executing "HeadObject" on "https://test-team.s3.amazonaws.com/data/m228/2019/06/01/1708092c8dfb526ea445452eb1be25357e95e2c5"; AWS HTTP error: Client error: `HEAD https://test-team.s3.amazonaws.com/data/m228/2019/06/01/1708092c8dfb526ea445452eb1be25357e95e2c5` resulted in a `404 Not Found` response NotFound (client): 404 Not Found (Request-ID: A36CB56A0CD87578) - '; private ${Exception}string = ''; protected $code = 0; protected $file = '/app/vendor/aws/aws-sdk-php/src/WrappedHttpHandler.php'; protected $line = 191; private ${Exception}trace = array (0 => array (...), 1 => array (...), 2 => array (...), 3 => array (...), 4 => array (...), 5 => array (...), 6 => array (...), 7 => array (...), 8 => array (...), 9 => array (...), 10 => array (...), 11 => array (...), 12 => array (...), 13 => array (...), 14 => array (...), 15 => array (...), 16 => array (...), 17 => array (...), 18 => array (...), 19 => array (...), 20 => array (...), 21 => array (...), 22 => array (...), 23 => array (...), 24 => array (...), 25 => array (...), 26 => array (...), 27 => array (...), 28 => array (...)); private ${Exception}previous = class GuzzleHttp\Exception\ClientException { private ${GuzzleHttp\Exception\RequestException}request = class GuzzleHttp\Psr7\Request { ... }; private ${GuzzleHttp\Exception\RequestException}response = class GuzzleHttp\Psr7\Response { ... }; private ${GuzzleHttp\Exception\RequestException}handlerContext = array (...); protected $message = 'Client error: `HEAD https://test-team.s3.amazonaws.com/data/m228/2019/06/01/1708092c8dfb526ea445452eb1be25357e95e2c5` resulted in a `404 Not Found` response'; private ${Exception}string = ''; protected $code = 404; protected $file = '/app/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php'; protected $line = 113; private ${Exception}trace = array (...); private ${Exception}previous = NULL; public $xdebug_message = '\nGuzzleHttp\\Exception\\ClientException: Client error: `HEAD https://test-team.s3.amazonaws.com/data/m228/2019/06/01/1708092c8dfb526ea445452eb1be25357e95e2c5` resulted in a `404 Not Found` response in /app/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php on line 113\n\nCall Stack:\n    0.2006     383568   1. {main}() /app/apps/web/public/index.php:0\n    0.2593    2669696   2. TDM\\Web\\Bootstrap\\Web->run() /app/apps/web/public/index.php:10\n    0.2593    2669696   3. Phalcon\\Mvc\\Application-...' }; private ${Aws\Exception\AwsException}monitoringEvents = array () }
  $unwrap = TRUE

{"code":2000,"status":"success","message":"","actions":[],"payload":[""]}
```

**Comment by niden** [comment]
@frankdejonge any thoughts on the above?

**Comment by frankdejonge** [comment]
@niden this is so weird. If I look at the call stack then it shows it originates from `checkExistenceWithCommand` which should catch the `S3Exception`. So this doesn't make any sense. I do see that you're using Phalcon. Could there be some sort of race-condition? Could you check your composer.lock for what versions of the aws-sdk you have installed? Also, which version of flysystem and the adapter, and what version of PHP do you run this in?

**Comment by niden** [comment]
@frankdejonge Thanks for the reply.

I don't think it is a race condition with Phalcon. The way we delete files is using an AJAX call from the front end and the `delete` is the only thing that is happening in that method/action

Versions
```
php 7.1.30

league/flysystem - 1.0.46
league/flysystem-aws-s3-v3 - 1.0.21
aws/aws-sdk-php - 3.71.4
```

**Comment by niden** [comment]
@frankdejonge An update on this.

The issue comes from `deleteObject`. Flysystem calls it and deletes the object and then does a `doesObjectExist` and if it exists that means that the record has not been deleted, if not, successful delete. 

I have stripped everything down and have replicated the functionality in a simple class and the problem appeared. I looked at the AWS documentation and it appears that `deleteObject` should return a flag (`true/false`) in the `Aws\Result` object returned. However that flag is always false, despite the fact that the object does get deleted indeed.

So I cannot find out whether the file has been correctly deleted by looking at the result returned from `deleteObject` since I am always getting `false` back. 

If I do what flysystem does and issue a `has` (or `doesObjectExist`) that invokes a promise and that is where everything falls apart again.

Finally I issued ['http' => ['synchronous' => true]]` in the parameters of the S3 client but still that did not help.

For us the workaround is to always return true for delete. We will also introduce a script that will check the database and the bucket and if the object does not exist in the database it will delete it, but that will be a CLI task.

I am closing this and hoping that this can help someone.
